### ISSUE TITLE
Question about face3dproj_forward

### ISSUE THREAD

**Issue by xczhao117** [tfwu/FaceDetection-ConvNet-3D#4]
my question is mainly about this function:

		MSHADOW_XINLINE DType Eval(index_t i, index_t j) const {
			using namespace std;
			const index_t dim = j % 2;
			const index_t nk = j / 2;
			const index_t w = i % src_width_;
			const index_t h = (i / src_width_) % src_height_;
			const index_t n = i / src_width_ / src_height_;

			DType a0 = data_src_.Eval(n * num_parameters_ * src_height_ +
									  (0 + dim * 4) * src_height_ + h, w);
			DType a1 = data_src_.Eval(n * num_parameters_ * src_height_ + 
									  (1 + dim * 4) * src_height_ + h, w);
			DType a2 = data_src_.Eval(n * num_parameters_ * src_height_ +
									  (2 + dim * 4) * src_height_ + h, w);
			DType a3 = data_src_.Eval(n * num_parameters_ * src_height_ +
									  (3 + dim * 4) * src_height_ + h, w);

			DType x0 = mface_src_.Eval(nk, 0);
			DType x1 = mface_src_.Eval(nk, 1);
			DType x2 = mface_src_.Eval(nk, 2);

			return a0 * x0 + a1 * x1 + a2 * x2 + a3 + \
				   (dim == 0 ? static_cast<DType>(h) / spatial_scale_ :
							   static_cast<DType>(w) / spatial_scale_);
		}

Let's set the input data shape to be (1,3,300,300), then the shape of the input data of "face3d_proj" layer should be (1,8,152,152). And the shape of the output is (1,152,152,20).

I guess the authors used the 8 parameters and the mean face to predict 10 keypoints for each point in 152x152 map.

I am not familiar with mshadow, could anyone explain how the function Eval posted above works, especially how to map the index (dim, nk, w h, n)?
Thx!

**Comment by YunzhuLi** [comment]
Assume the input of `face3d_proj` is of size `batch_size * num_parameters_ * src_height_ * src_width_`.

Here is the meaning of the output index:
`n` index of the input image in the batch
`h` row index of target position
`w` column index of target position
`nk` index of the corresponding keypoint
`dim` 0 if we are dealing with the row index of the corresponding keypoint, otherwise 1
We combined `nk` and `dim` into one dimension, resulting in a 4D output of size `batch_size * src_height_ * src_width_ * (num_keypoints_ * 2)`. Here `num_parameters_` is 8 and `num_keypoints_` is 10 referring from our paper.

The code here is just 
1) extracting parameters and coordinates from input tensor and pre-defined mean face individually,
2) calculating the transformation as we have stated in the paper, 
3) adding a shift to calculate the absolute position in the image, then
4) feeding the result to the output.
